libclang_setup.py should allow for customizable locations of libclang

From feedback over at https://github.com/RobotLocomotion/drake/pull/14648#issuecomment-779499081, @jamiesnape suggested that the currently hardcoded path for libclang might be able to be parameterized. Currently the executable that calls it `mkdoc` doesn't accept any command-line arguments, but it seems like some toolchain configurations from bazel might be able to be passed on to mkdoc. 
 
https://github.com/RobotLocomotion/drake/blob/6f08f1b83b88a7932ab162f0412e820bddd4a580/tools/workspace/pybind11/libclang_setup.py#L40

I ran into this issue when trying to build with clang-11 and c++20 (https://github.com/RobotLocomotion/drake/pull/14648). The `mkdoc` targets won't make use of the CC variable here.
```
bazel build --cxxopt=-std=c++20 --repo_env=CC=clang-11  //...
```

@jwnimmer-tri Pretty much. It complains that c++20 is not a valid c++ value.

```
error: invalid value 'c++20' in '-std=c++20'
note: use 'c++98' or 'c++03' for 'ISO C++ 1998 with amendments' standard
note: use 'gnu++98' or 'gnu++03' for 'ISO C++ 1998 with amendments and GNU extensions' standard
note: use 'c++11' for 'ISO C++ 2011 with amendments' standard
note: use 'gnu++11' for 'ISO C++ 2011 with amendments and GNU extensions' standard
note: use 'c++14' for 'ISO C++ 2014 with amendments' standard
note: use 'gnu++14' for 'ISO C++ 2014 with amendments and GNU extensions' standard
note: use 'c++17' for 'ISO C++ 2017 with amendments' standard
note: use 'gnu++17' for 'ISO C++ 2017 with amendments and GNU extensions' standard
note: use 'c++2a' for 'Working draft for ISO C++ 2020' standard
note: use 'gnu++2a' for 'Working draft for ISO C++ 2020 with GNU extensions' standard
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/hash.h:138:39: error: no template named 'optional' in namespace 'std'
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_monomial_basis_element.h:168:24: error: no member named 'is_same_v' in namespace 'std'
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_monomial_basis_element.h:168:34: error: 'BasisElement' does not refer to a value
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_monomial_basis_element.h:166:22: note: declared here
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_monomial_basis_element.h:172:24: error: no member named 'is_same_v' in namespace 'std'
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_monomial_basis_element.h:172:34: error: 'BasisElement' does not refer to a value
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_monomial_basis_element.h:166:22: note: declared here
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_generic_polynomial.h:157:14: error: no member named 'is_same_v' in namespace 'std'
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_generic_polynomial.h:157:48: error: expected '(' for function-style cast or type construction
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_generic_polynomial.h:288:10: error: no member named 'is_base_of_v' in namespace 'std'
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_generic_polynomial.h:288:23: error: 'PolynomialBasisElement' does not refer to a value
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_polynomial_basis_element.h:46:7: note: declared here
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_generic_polynomial.h:288:60: error: expected ';' after alias declaration
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_generic_polynomial.h:289:23: error: use of undeclared identifier 'BasisElement'
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_generic_polynomial.h:292:1: error: no template named 'GenericPolynomialEnable'
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_generic_polynomial.h:298:1: error: no template named 'GenericPolynomialEnable'
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_generic_polynomial.h:305:1: error: no template named 'GenericPolynomialEnable'
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_generic_polynomial.h:311:1: error: no template named 'GenericPolynomialEnable'
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_generic_polynomial.h:317:1: error: no template named 'GenericPolynomialEnable'
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_generic_polynomial.h:323:1: error: no template named 'GenericPolynomialEnable'
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_generic_polynomial.h:329:1: error: no template named 'GenericPolynomialEnable'
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_headers/drake/common/symbolic_generic_polynomial.h:335:1: error: no template named 'GenericPolynomialEnable'
fatal error: too many errors emitted, stopping now [-ferror-limit=]
Traceback (most recent call last):
  File "/home/brawner/.cache/bazel/_bazel_brawner/9e7b33f8778394ea8277807e1cb12d27/sandbox/linux-sandbox/1380/execroot/drake/bazel-out/host/bin/tools/workspace/pybind11/mkdoc.runfiles/drake/tools/workspace/pybind11/mkdoc.py", line 841, in <module>
    main()
  File "/home/brawner/.cache/bazel/_bazel_brawner/9e7b33f8778394ea8277807e1cb12d27/sandbox/linux-sandbox/1380/execroot/drake/bazel-out/host/bin/tools/workspace/pybind11/mkdoc.runfiles/drake/tools/workspace/pybind11/mkdoc.py", line 801, in main
    raise RuntimeError(
RuntimeError: Parsing headers using the clang library failed with 20 error(s) and 1 fatal error(s)
----------------
Note: The failure of target //tools/workspace/pybind11:mkdoc (with exit code 1) may have been caused by the fact that it is running under Python 3 instead of Python 2. Examine the error to determine if that appears to be the problem. Since this target is built in the host configuration, the only way to change its version is to set --host_force_python=PY2, which affects the entire build.

If this error started occurring in Bazel 0.27 and later, it may be because the Python toolchain now enforces that targets analyzed as PY2 and PY3 run under a Python 2 and Python 3 interpreter, respectively. See https://github.com/bazelbuild/bazel/issues/7899 for more information.
----------------
```

--- a/tools/skylark/pybind.bzl
+++ b/tools/skylark/pybind.bzl
@@ -368,7 +368,7 @@ def _generate_pybind_documentation_header_impl(ctx):
              ",".join(ctx.attr.ignore_dirs_for_coverage))
     for p in ctx.attr.exclude_hdr_patterns:
         args.add("-exclude-hdr-patterns=" + p)
-    args.add_all(ctx.fragments.cpp.cxxopts, uniquify = True)
+    args.add("-std=c++17")
     args.add("-w")
 
     # N.B. This is for `targets` only.
@@ -415,7 +415,6 @@ generate_pybind_documentation_header = rule(
             default = [],
         ),
     },
-    fragments = ["cpp"],
     implementation = _generate_pybind_documentation_header_impl,
     output_to_genfiles = True,
 )

